// eslint-config-mi9/index.js
module.exports = {
  // Ensure ESLint can parse ES2016+ code
  parser: require.resolve('babel-eslint'),

  extends: 'eslint-config-airbnb',

  // Add Mocha global variables
  env: {
    mocha: true,
  },

  // Enable rules for mocha
  plugins: [
    'mocha',
  ],

  // Add some global constants used in our projects
  globals: {
    __DEVELOPMENT__: true,
    __CLIENT__: true,
    __SERVER__: true,
    __DEVTOOLS__: true,

    navigator: true,
    window: true,
    document: true,
    requestAnimationFrame: true,
    cancelAnimationFrame: true,
    fetch: true,
    CustomEvent: true,
    sessionStorage: true,
  },

  rules: {
    // Don't warn about anonymous functions
    'func-names': 'off',
    // Discourage looooong lines
    'max-len': ['warn', 120],
    // disallow un-paren'd mixes of different operators
    'no-mixed-operators': ['error', {
      groups: [
        ['&', '|', '^', '~', '<<', '>>', '>>>'],
        ['==', '!=', '===', '!==', '>', '>=', '<', '<='],
        ['&&', '||'],
        ['in', 'instanceof'],
      ],
      allowSamePrecedence: false,
    }],
    // Don't disallow variable names that start or end with underscore
    'no-underscore-dangle': 'off',
    // Allow using foo && bar() as a shorthand to conditionally run a function
    'no-unused-expressions': ['error', { allowShortCircuit: true, allowTernary: true }],
    // Discourage TODO and FIXME comments
    'no-warning-comments': 'warn',

    // @see https://github.com/lo1tuma/eslint-plugin-mocha/tree/master/docs/rules
    'mocha/handle-done-callback': 'error',
    'mocha/no-exclusive-tests': 'error',
    'mocha/no-global-tests': 'error',
    'mocha/no-identical-title': 'error',
    'mocha/no-pending-tests': 'error',
    'mocha/no-sibling-hooks': 'error',
    'mocha/no-return-and-callback': 'error',
    'mocha/no-top-level-hooks': 'error',

    // We don't name our components with .jsx
    'react/jsx-filename-extension': 'off',
  },
};
